### ISSUE TITLE
Predict does not work with TfmdLists

### ISSUE THREAD

**Issue by borisdayma** [fastai/fastai2#102]
When creating a learner from TfmdLists, `predict` does not work.

Here we create a U-net where output = input (as `TensorImage`)

```python
from fastai2.vision.all import *

path = untar_data(URLs.MNIST)
items = get_image_files(path)

class Duplicate(Transform):
    order=10
    def encodes(self, o):
        return (o,o)

dsrc = TfmdLists(items, tfms=[PILImage.create, Resize(128), ToTensor, Duplicate])
dls = dsrc.dataloaders(bs=4, after_batch=[IntToFloatTensor])

learn = unet_learner(dls=dls, arch=resnet18, n_out=3, loss_func=MSELossFlat())
learn.predict(items[0])
```

Error:

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-7-2ccff2aa1d8f> in <module>
----> 1 learn.predict(items[0])

~/Projects/fastai2/fastai2/learner.py in predict(self, item, rm_type_tfms, with_input)
    331     def predict(self, item, rm_type_tfms=None, with_input=False):
    332         dl = self.dls.test_dl([item], rm_type_tfms=rm_type_tfms)
--> 333         inp,preds,_,dec_preds = self.get_preds(dl=dl, with_input=True, with_decoded=True)
    334         dec = self.dls.decode_batch((*tuplify(inp),*tuplify(dec_preds)))[0]
    335         i = getattr(self.dls, 'n_inp', -1)

~/Projects/fastai2/fastai2/learner.py in get_preds(self, ds_idx, dl, with_input, with_decoded, with_loss, act, **kwargs)
    319             for mgr in ctx_mgrs: stack.enter_context(mgr)
    320             self(_before_epoch)
--> 321             self._do_epoch_validate(dl=dl)
    322             self(_after_epoch)
    323             if act is None: act = getattr(self.loss_func, 'activation', noop)

~/Projects/fastai2/fastai2/learner.py in _do_epoch_validate(self, ds_idx, dl)
    278             dl,old,has = change_attrs(dl, names, [False,False])
    279             self.dl = dl;                                    self('begin_validate')
--> 280             with torch.no_grad(): self.all_batches()
    281         except CancelValidException:                         self('after_cancel_validate')
    282         finally:

~/Projects/fastai2/fastai2/learner.py in all_batches(self)
    246     def all_batches(self):
    247         self.n_iter = len(self.dl)
--> 248         for o in enumerate(self.dl): self.one_batch(*o)
    249 
    250     def one_batch(self, i, b):

~/Projects/fastai2/fastai2/data/load.py in __iter__(self)
     95         self.randomize()
     96         self.before_iter()
---> 97         for b in _loaders[self.fake_l.num_workers==0](self.fake_l):
     98             if self.device is not None: b = to_device(b, self.device)
     99             yield self.after_batch(b)

~/.local/share/virtualenvs/l2ab-YhTBTBv9/lib/python3.7/site-packages/torch/utils/data/dataloader.py in __next__(self)
    817             else:
    818                 del self._task_info[idx]
--> 819                 return self._process_data(data)
    820 
    821     next = __next__  # Python 2 compatibility

~/.local/share/virtualenvs/l2ab-YhTBTBv9/lib/python3.7/site-packages/torch/utils/data/dataloader.py in _process_data(self, data)
    844         self._try_put_index()
    845         if isinstance(data, ExceptionWrapper):
--> 846             data.reraise()
    847         return data
    848 

~/.local/share/virtualenvs/l2ab-YhTBTBv9/lib/python3.7/site-packages/torch/_utils.py in reraise(self)
    383             # (https://bugs.python.org/issue2651), so we work around it.
    384             msg = KeyErrorMessage(msg)
--> 385         raise self.exc_type(msg)

TypeError: Caught TypeError in DataLoader worker process 0.
Original Traceback (most recent call last):
  File "/home/boris/.local/share/virtualenvs/l2ab-YhTBTBv9/lib/python3.7/site-packages/torch/utils/data/_utils/worker.py", line 178, in _worker_loop
    data = fetcher.fetch(index)
  File "/home/boris/.local/share/virtualenvs/l2ab-YhTBTBv9/lib/python3.7/site-packages/torch/utils/data/_utils/fetch.py", line 34, in fetch
    data = next(self.dataset_iter)
  File "/home/boris/Projects/fastai2/fastai2/data/load.py", line 106, in create_batches
    yield from map(self.do_batch, self.chunkify(res))
  File "/home/boris/Projects/fastai2/fastai2/data/load.py", line 127, in do_batch
    def do_batch(self, b): return self.retain(self.create_batch(self.before_batch(b)), b)
  File "/home/boris/Projects/fastai2/fastai2/data/load.py", line 126, in create_batch
    def create_batch(self, b): return (fa_collate,fa_convert)[self.prebatched](b)
  File "/home/boris/Projects/fastai2/fastai2/data/load.py", line 47, in fa_collate
    else default_collate(t))
  File "/home/boris/.local/share/virtualenvs/l2ab-YhTBTBv9/lib/python3.7/site-packages/torch/utils/data/_utils/collate.py", line 81, in default_collate
    raise TypeError(default_collate_err_msg_format.format(elem_type))
TypeError: default_collate: batch must contain tensors, numpy arrays, numbers, dicts or lists; found <class 'pathlib.PosixPath'>
```

I think issue #70 is directly related to this.

**Comment by sgugger** [comment]
Thanks for flagging, there was so custom code needed for when the datasets were built with TfmdLists. I added this and your example works now, could you check if it solved #70 too?

**Comment by borisdayma** [comment]
Yes let me check on this
